### PR TITLE
server: Add remaining support for simple copy/paste

### PR DESCRIPTION
**Pull request by soreau**
- Add function to set callback for client_cut_text
- Add server_cut_text structure and function

### CODE REVIEW

**Comment by any1** [comment]
Nice! The rest of the code has `{` on the same line as `if` and `while`. Can you fix that for this and the wayvnc PR also?

**Comment by any1** [comment]
Also, we don't use `{ }` when there's only a single line in the code block, so prefer
```
if (stuff)
    do_stuff();
```

**Comment by soreau** [comment]
> Nice! The rest of the code has `{` on the same line as `if` and `while`. Can you fix that for this and the wayvnc PR also?
> Also, we don't use `{ }` when there's only a single line in the code block, so prefer
> 
> ```
> if (stuff)
>     do_stuff();
> ```

These should be fixed now.

File: include/common.h [hunk 1]
```diff
@@ -73,6 +73,7 @@ struct nvnc_client {
 	int n_pending_requests;
 	bool is_updating;
 	nvnc_client_fn cleanup_fn;
+	nvnc_client_cut_fn client_cut_fn;
```
Comment (any1) [inline include/common.h:76]: What scenario do you have in mind where it's useful to be able to set a per-client cut text callback?

File: include/neatvnc.h [hunk 2]
```diff
@@ -49,6 +49,7 @@ typedef void (*nvnc_damage_fn)(struct pixman_region16* damage, void* userdata);
 typedef bool (*nvnc_auth_fn)(const char* username, const char* password,
                              void* userdata);
 typedef void (*nvnc_render_fn)(struct nvnc_display*, struct nvnc_fb*);
+typedef void (*nvnc_client_cut_fn)(struct nvnc_client*, char* data, uint32_t len);
```
Comment (any1) [inline include/neatvnc.h:52]: `data` should be `const`. The word "data" is such a generic word, would it not be more fitting to call this "text"? Why is there a length parameter? Can this not be a zero terminated string?

File: include/neatvnc.h [hunk 3]
```diff
@@ -70,6 +71,7 @@ void nvnc_set_pointer_fn(struct nvnc* self, nvnc_pointer_fn);
 void nvnc_set_fb_req_fn(struct nvnc* self, nvnc_fb_req_fn);
 void nvnc_set_new_client_fn(struct nvnc* self, nvnc_client_fn);
 void nvnc_set_client_cleanup_fn(struct nvnc_client* self, nvnc_client_fn fn);
+void nvnc_set_data_transfer_fn(struct nvnc* self, nvnc_client_cut_fn fn);
```
Comment (any1) [inline include/neatvnc.h:74]: Why not `nvnc_set_client_cut_fn`?

File: include/neatvnc.h [hunk 4]
```diff
@@ -111,3 +113,6 @@ void nvnc_display_damage_whole(struct nvnc_display*);
 int nvnc_check_damage(struct nvnc_fb* fb0, struct nvnc_fb* fb1,
                       int x_hint, int y_hint, int width_hint, int height_hint,
                       nvnc_damage_fn on_check_done, void* userdata);
+
+void nvnc_server_cut_text(struct nvnc_display*, char *text, uint32_t length);
+void _nvnc_server_cut_text(struct nvnc_display*, char *text, uint32_t length);
```
Comment (any1) [inline include/neatvnc.h:118]: text should be const and the star should be on the left side of the space. Also, does there need to be length? `_nvnc_server_cut_text` should not be here as it is not a public API function.

File: include/rfb-proto.h [hunk 5]
```diff
@@ -144,7 +144,14 @@ struct rfb_client_cut_text_msg {
 	uint8_t type;
 	uint8_t padding[3];
 	uint32_t length;
-	char test[0];
+	char data[0];
```
Comment (any1) [inline include/rfb-proto.h:147]: Same as before, I prefer the word "text".

File: include/rfb-proto.h [hunk 6]
```diff
@@ -144,7 +144,14 @@ struct rfb_client_cut_text_msg {
...
+} RFB_PACKED;
+
+struct rfb_server_cut_text_msg {
```
Comment (any1) [inline include/rfb-proto.h:150]: These structs are the same, right? Why not just have one of them called `rfb_cut_text_msg`?

File: src/server.c [hunk 7]
```diff
@@ -584,6 +584,24 @@ static int on_client_pointer_event(struct nvnc_client* client)
 	return sizeof(*msg);
 }
 
+void _nvnc_server_cut_text(struct nvnc_display* display, char* text, uint32_t len)
```
Comment (any1) [inline src/server.c:587]: Why per display? Should this not be per server?

File: src/server.c [hunk 8]
```diff
@@ -584,6 +584,24 @@ static int on_client_pointer_event(struct nvnc_client* client)
...
+{
+	size_t size = sizeof(struct rfb_server_cut_text_msg);
+
+	struct rfb_server_cut_text_msg* msg = calloc(1, size);
```
Comment (any1) [inline src/server.c:591]: This does not need to be heap allocated.

File: include/neatvnc.h (hunk 2 above)
Comment (any1) [reply include/neatvnc.h:52]: Also, let's call the type `nvnc_cut_text_fn`.

File: include/neatvnc.h (hunk 4 above)
Comment (any1) [reply include/neatvnc.h:118]: Also, I'd prefer the name `nvnc_send_cut_text`. The names "client cut text" and "server cut text" are implementation details leaking from the protocol.

**Review by any1 — CHANGES_REQUESTED** [review]

File: src/server.c [hunk 9]
```diff
@@ -584,18 +584,39 @@ static int on_client_pointer_event(struct nvnc_client* client)
...
 }
 
+EXPORT
+void nvnc_send_cut_text(struct nvnc* server, const char* text)
+{
+	size_t len = strlen(text);
+
+	struct rfb_cut_text_msg msg;
+
+	msg.type = 3;
```
Comment (any1) [inline src/server.c:594]: This should be `RFB_SERVER_TO_CLIENT_SERVER_CUT_TEXT`

File: src/server.c [hunk 10]
```diff
@@ -584,18 +584,39 @@ static int on_client_pointer_event(struct nvnc_client* client)
...
+	msg.length = htonl(len);
+
+	struct nvnc_client *client, *tmp;
+	LIST_FOREACH_SAFE (client, &server->clients, link, tmp) {
```
Comment (any1) [inline src/server.c:598]: `LIST_FOREACH` will do.

File: include/common.h [hunk 11]
```diff
@@ -96,6 +96,7 @@ struct nvnc {
 	nvnc_pointer_fn pointer_fn;
 	nvnc_fb_req_fn fb_req_fn;
 	nvnc_client_fn new_client_fn;
+	nvnc_cut_text_fn client_cut_fn;
```
Comment (any1) [inline include/common.h:99]: Let's just call this `cut_text_fn`

File: include/neatvnc.h [hunk 12]
```diff
@@ -70,6 +71,7 @@ void nvnc_set_pointer_fn(struct nvnc* self, nvnc_pointer_fn);
 void nvnc_set_fb_req_fn(struct nvnc* self, nvnc_fb_req_fn);
 void nvnc_set_new_client_fn(struct nvnc* self, nvnc_client_fn);
 void nvnc_set_client_cleanup_fn(struct nvnc_client* self, nvnc_client_fn fn);
+void nvnc_set_client_cut_fn(struct nvnc* self, nvnc_cut_text_fn fn);
```
Comment (any1) [inline include/neatvnc.h:74]: Maybe call this `nvnc_set_cut_text_receive_fn`?

**Review by any1 — CHANGES_REQUESTED** [review]
The commit message says "fixes #4", but the correct thing to say is "closes #4". ;)

**Comment by soreau** [comment]
The length argument is necessary because we're not always guaranteed to be dealing with null terminated strings.

**Review by any1 — APPROVED** [review]

**Comment by any1** [comment]
Thanks!

**Pull request merged**